fuel-vm/src/interpreter/executors/main.rs

```diff
@@ -48,46 +50,121 @@ impl<T> Interpreter<PredicateStorage, T> {
     ) -> Result<PredicatesChecked, PredicateVerificationFailed>
     where
         Tx: ExecutableTransaction,
-        <Tx as IntoChecked>::Metadata: CheckedMetadata,
+        <Tx as IntoChecked>::CheckedMetadata: CheckedMetadata,
     {
         if !checked.transaction().check_predicate_owners(&params) {
             return Err(PredicateVerificationFailed::InvalidOwner);
         }
 
         let mut vm = Interpreter::with_storage(PredicateStorage::default(), params, gas_costs);
 
-        // Needed for now because checked is only freed once the value is collected into a Vec
-        #[allow(clippy::needless_collect)]
-        let predicates: Vec<_> = (0..checked.transaction().inputs().len())
-            .filter_map(|i| RuntimePredicate::from_tx(&params, checked.transaction(), i))
-            .collect();
+        let mut cumulative_gas_used: Word = 0;
```

Comment:
It seems you don't need to change the old logic. You can return it back, and at the end of the loop add:

```rust
 if vm.registers[RegId::GGAS] != 0 {
     return Err(PredicateVerificationFailed::GasMismatch);
 }
 cumulative_gas_used = cumulative_gas_used
     .checked_add(gas_used)
     .ok_or_else(|| PredicateVerificationFailed::OutOfGas)?;
```

Reply:
the RuntimePredicate returned here does not have the gas_used returned, only the range of the predicate itself, this complexity led to the change in the logic to more closely match the iteration similar to estimate_predicates

Reply:
are you commenting on the removal of the predicate collection in favor of iterating over all inputs or simply the cumulative_gas_used logic?